types/wlr_output.c

```diff
@@ -172,24 +172,24 @@ bool wlr_output_set_cursor(struct wlr_output *output,
 
 bool wlr_output_set_cursor_surface(struct wlr_output *output,
 		struct wlr_surface *surface, int32_t hotspot_x, int32_t hotspot_y) {
-	if (output->impl->set_cursor) {
-		output->impl->set_cursor(output, NULL, 0, 0, 0, 0, 0);
-		wlr_log(L_INFO, "TODO: wlr_output_set_cursor_surface for hw cursors");
+	struct wl_shm_buffer *buffer = wl_shm_buffer_get(surface->current->buffer);
+	if (buffer == NULL) {
+		return false;
 	}
 
-	output->cursor.is_sw = true;
-	output->cursor.width = surface->current->width;
-	output->cursor.height = surface->current->height;
-	output->cursor.hotspot_x = hotspot_x;
-	output->cursor.hotspot_y = hotspot_y;
-
-	wlr_texture_destroy(output->cursor.texture);
-	output->cursor.texture = surface->texture;
-
-	wlr_renderer_destroy(output->cursor.renderer);
-	output->cursor.renderer = surface->renderer;
+	uint32_t format = wl_shm_buffer_get_format(buffer);
+	if (format != WL_SHM_FORMAT_ARGB8888) {
+		return false;
+	}
 
-	return true;
+	void *data = wl_shm_buffer_get_data(buffer);
+	int32_t width = wl_shm_buffer_get_width(buffer);
+	int32_t height = wl_shm_buffer_get_height(buffer);
+	wl_shm_buffer_begin_access(buffer);
+	bool ok = wlr_output_set_cursor(output, data, width, width, height,
```

Comment:
You need to pass the stride properly.

Reply:
Since we only support `ARGB8888`, this isn't a big deal. Also, `wl_shm_buffer_get_stride` returns the stride in bytes, and we need to pass it in pixels (so is it really useful?). See https://github.com/swaywm/wlroots/blob/master/backend/wayland/output.c#L61

Reply:
The stride has to do with the distance between different rows. There is no guarantee that stride == width (i.e. tightly packed).

Reply:
Hmmm, you're right, I haven't thought about that. I've seen this done somewhere else in the code, I guess I'll change that too if possible.